pkg/controllermanager/controller/shoot/shoot_hibernation_control.go

```diff
@@ -56,85 +60,144 @@ func (c *Controller) shootHibernationUpdate(oldObj, newObj interface{}) {
 		newSchedule = getShootHibernationSchedules(newShoot)
 	)
 
-	if !reflect.DeepEqual(oldSchedule, newSchedule) {
+	if !reflect.DeepEqual(oldSchedule, newSchedule) && len(newSchedule) > 0 {
 		key, err := cache.MetaNamespaceKeyFunc(newObj)
 		if err != nil {
 			gardenlogger.Logger.Errorf("Couldn't get key for object %+v: %v", newObj, err)
 			return
 		}
-		c.shootHibernationQueue.Add(key)
+		parsedSchedules, err := parseHibernationSchedules(newSchedule)
+		if err != nil {
+			gardenlogger.Logger.Infof("Could not parse hibernation schedules for shoot %s: %v", client.ObjectKeyFromObject(newShoot), err)
+			return
+		}
+		requeueAfter := nextHibernationTimeDuration(parsedSchedules, TimeNow())
+		c.shootHibernationQueue.AddAfter(key, requeueAfter)
 	}
 }
 
-func (c *Controller) shootHibernationDelete(obj interface{}) {
-	shoot, ok := obj.(*gardencorev1beta1.Shoot)
-	if !ok {
-		return
+// ControllerName is the name of the controller.
+const ControllerName = "hibernation"
+
+// TimeNow returns the current time. Exposed for testing.
+var TimeNow = time.Now
+
+// ParsedHibernationSchedule holds the loaded location, parsed cron schedule and information whether
+// the cluster should be hibernated or woken up.
+type ParsedHibernationSchedule struct {
+	Location *time.Location
+	Schedule cron.Schedule
+	Enabled  bool
+}
+
+// Next returns the next activation time in UTC, later than the given time.
+// The given time is converted in the schedule's location.
+func (s *ParsedHibernationSchedule) Next(t time.Time) time.Time {
+	return s.Schedule.Next(t.In(s.Location)).UTC()
+}
+
+// Prev returns the previous activation time in UTC that is between the two given times.
+// The times are converted in the schedule's location.
+func (s *ParsedHibernationSchedule) Prev(from, to time.Time) *time.Time {
+	t1 := s.Schedule.Next(from.In(s.Location))
+	if t1.After(to) {
+		return nil
 	}
 
-	if shootHasHibernationSchedules(shoot) {
-		key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
-		if err != nil {
-			gardenlogger.Logger.Errorf("Couldn't get key for object %+v: %v", obj, err)
-			return
+	for {
+		t2 := s.Schedule.Next(t1)
+		if t2.After(to) {
+			break
 		}
-		c.shootHibernationQueue.Add(key)
+		t1 = t2
 	}
+	inUTC := t1.UTC()
+	return &inUTC
```

Comment:
Why is the first `Next` call "special" (not part of the `for` loop)? Can't this be combined?

```golang
t := from.In(s.Location)

for {
    next := s.Schedule.Next(t)
    if next.UTC().After(to.UTC()) {
        break
    }
    t = next
}

inUTC := t.UTC()
return &inUTC
```

Reply:
Like this, the method will return `from` if the previous execution time in the schedule was before `from`. 
E.g. `from` is 14:00:00, `to` is 14:30:00 and the schedule is configured to run everyday at 13:30:00, then we would want to return nil from the method instead of 14:00:00.

 I couldn't come up with a way to avoid the virst `Next()` call so that the function always returns a time that is after `from` and before `to`

Reply:
OK I figured out how to do this entire code on 3 lines :)